tests/components/numato/test_numato.py

```diff
@@ -0,0 +1,325 @@
+    assert not numato_fixture.devices[0].values[6]
+    await async_turn_off(hass, "switch.numato_switch_mock_port5")
+    assert not numato_fixture.devices[0].values[5]
+    await async_turn_off(hass, "switch.numato_switch_mock_port6")
+    assert not numato_fixture.devices[0].values[6]
+
+
+@pytest.mark.asyncio
+async def test_failing_async_added_to_hass(hass, numato_fixture, monkeypatch):
+    """Test condition when a sensor update fails."""
+    monkeypatch.setattr(numato_fixture.NumatoDeviceMock, "setup", mockup_raise)
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    await hass.async_block_till_done()  # wait until services are registered
+
+
+@pytest.mark.asyncio
+async def test_failing_sensor_update(hass, numato_fixture, monkeypatch):
+    """Test condition when a sensor update fails."""
+    monkeypatch.setattr(numato_fixture.NumatoDeviceMock, "adc_read", mockup_raise)
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    await hass.async_block_till_done()  # wait until services are registered
+
+
+@pytest.mark.asyncio
+async def test_hass_numato_api(hass, numato_fixture):
+    """Test regular device access."""
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    api = hass.data["numato"]["api"]
+    # regular operations
+    api.read_adc_input(0, 1)
+    api.read_input(0, 2)
+    api.write_output(0, 2, 1)
+
+    def gen_callback(expected_port, expected_value):
+        def cb_mockup(port, value):
+            assert port == expected_port
+            assert value == expected_value
+
+        return cb_mockup
+
+    api.edge_detect(0, 2, gen_callback(2, 1))
+    numato_fixture.devices[0].mockup_inject_notification(2, 1)
+    api.edge_detect(0, 2, gen_callback(2, 0))
+    numato_fixture.devices[0].mockup_inject_notification(2, 0)
+
+
+@pytest.mark.asyncio
+async def test_hass_numato_api_irregular_unhandled(hass, numato_fixture):
+    """Test irregular, unhandled operations.
+
+    Establishes that these don't throw from numato-gpio or are handled in the
+    numato component's API.
+    """
+    assert await async_setup_component(hass, "numato", NUMATO_CFG)
+    api = hass.data["numato"]["api"]
+    api.read_adc_input(0, 5)  # adc_read from output
+    api.read_input(0, 6)  # read from output
+    api.write_output(0, 2, 1)  # write to input
+
+
+@pytest.mark.asyncio
+async def test_hass_numato_api_errors(hass, numato_fixture, monkeypatch):
+    """Test whether Home Assistant numato API (re-)raises errors."""
+    numato_fixture.discover()
+    monkeypatch.setattr(numato_fixture.devices[0], "setup", mockup_raise)
+    monkeypatch.setattr(numato_fixture.devices[0], "adc_read", mockup_raise)
+    monkeypatch.setattr(numato_fixture.devices[0], "read", mockup_raise)
+    monkeypatch.setattr(numato_fixture.devices[0], "write", mockup_raise)
+    api = numato.NumatoAPI()
+    with pytest.raises(NumatoGpioError):
+        api.setup_input(0, 5)
```

Comment:
Incorporate these tests in the platform tests instead and check that the entity handles errors correctly and results in the expected state. Ie we don't access `api` directly, just assert the state of the entity via the state machine.

Reply:
I think the tests are already covered by the `regular_hass_operations` test. I just thought in case of a bad change of the `NumatoAPI` code the root cause could be identified quicker since the `regular_hass_operations` covers quite some code which leaves no very specific clue to where the problem is. I'll just remove it if you see no value, or find it unconventional.

Reply:
Ok, please remove it then. We want the tests for integrations to be agnostic to integration details. That makes the tests more robust.

Reply:
done.